samples/simulation/qaoa/QAOA.qs

```diff
@@ -0,0 +1,240 @@
+namespace Microsoft.Quantum.Samples.QAOA {
+    open Microsoft.Quantum.Intrinsic;
+    open Microsoft.Quantum.Canon;
+    open Microsoft.Quantum.Measurement;
+    open Microsoft.Quantum.Convert;
+    open Microsoft.Quantum.Arrays;
+    open Microsoft.Quantum.Math;
+    open Microsoft.Quantum.Diagnostics;
+
+    /// # Summary
+    /// This operation applies the X-rotation to each qubit. We can think of it as time 
+    /// evolution induced by applying H = - \sum_i X_i for time t.
+    ///
+    /// # Input
+    /// ## target
+    /// Target qubit register
+    /// ## time
+    /// Time passed in evolution of X rotation
+    operation DriverHamiltonian(target: Qubit[], time: Double) : Unit {
+        for(qubit in target)
+        {
+            Rx(-2.0 * time, qubit);
+        }
+    }
+
+    /// # Summary
+    /// This applies the Z-rotation according to the instance Hamiltonian. 
+    /// We can think of it as Hamiltonian time evolution for time t induced
+    /// by the Ising Hamiltonian $\sum_ij J_ij Z_i Z_j + \sum_i h_i Z_i$.
+    ///
+    /// # Input
+    /// ## target
+    /// Qubit register that encodes the Spin values in the Ising Hamiltonian
+    /// ## time
+    /// Time point in evolution
+    /// ## weights
+    /// Ising magnetic field or "weights" encoding the constraints of our
+    /// traveling Santa problem.
+    /// ## coupling
+    /// Ising coupling term or "penalty" encoding the constraints of our
+    /// traveling Santa problem.
+    operation InstanceHamiltonian(
+        target: Qubit[], 
+        time: Double, 
+        weights: Double[], 
+        coupling: Double[]
+    ) : Unit {
+        using (auxiliary = Qubit())
+        {
+            for((h, qubit) in Zip(weights, target))
+            {
+                Rz(2.0 * time * h, qubit);
+            }
+            for(i in 0..5)
+            {
+                for (j in i+1..5)
+                {
+                    CNOT(target[i], auxiliary);
+                    CNOT(target[j], auxiliary);
+                    Rz(2.0 * time * coupling[6 * i + j], auxiliary);
+                    CNOT(target[i], auxiliary);
+                    CNOT(target[j], auxiliary);
```

Comment:
Can this be rewritten to use `within`/`apply`, so as to demonstrate that feature? I think the two CNOTs should commute, such that the following should work:
```suggestion
                    within {
                        CNOT(target[i], auxiliary);
                        CNOT(target[j], auxiliary);
                    } apply {
                        Rz(2.0 * time * coupling[6 * i + j], auxiliary);
                    }
```